bundle/Command/TagContentByTypesCommand.php

```diff
@@ -0,0 +1,195 @@
+use Netgen\TagsBundle\API\Repository\Values\Tags\Tag;
+use Netgen\TagsBundle\Core\FieldType\Tags\Value as TagFieldValue;
+use Symfony\Component\Console\Command\Command;
+use Symfony\Component\Console\Exception\InvalidOptionException;
+use Symfony\Component\Console\Input\InputInterface;
+use Symfony\Component\Console\Input\InputOption;
+use Symfony\Component\Console\Output\OutputInterface;
+use Symfony\Component\Console\Style\SymfonyStyle;
+use Throwable;
+
+use function array_filter;
+use function array_map;
+use function array_unique;
+use function array_values;
+use function explode;
+use function sprintf;
+use function trim;
+
+final class TagContentByTypesCommand extends Command
+{
+    private SymfonyStyle $style;
+    private InputInterface $input;
+
+    public function __construct(
+        private Repository              $repository,
+        private TagsService             $tagsService,
+        private ConfigResolverInterface $configResolver,
+    ) {
+        // Parent constructor call is mandatory for commands registered as services
+        parent::__construct();
+    }
+
+    protected function configure(): void
+    {
+        $this->setDescription('Add tags to content')
+            ->addOption('parent-location', null, InputOption::VALUE_REQUIRED)
+            ->addOption('content-types', null, InputOption::VALUE_REQUIRED)
+            ->addOption('tag-id', null, InputOption::VALUE_REQUIRED)
+            ->addOption('field-identifier', null, InputOption::VALUE_OPTIONAL);
+    }
+
+    protected function initialize(InputInterface $input, OutputInterface $output): void
+    {
+        $this->input = $input;
+        $this->style = new SymfonyStyle($input, $output);
+    }
+
+    protected function execute(InputInterface $input, OutputInterface $output): int
+    {
+        $query = new Query();
+        $query->filter = new Criterion\LogicalAnd([
+            new Criterion\ParentLocationId($this->getParentLocation()),
+            new Criterion\ContentTypeIdentifier($this->getContentTypes()),
+        ]);
+
+        $batchSize = 50;
+
+        $searchResults = $this->repository->getSearchService()->findContent($query);
+        $totalResults = $searchResults->totalCount;
+
+        $this->style->newLine();
+        $this->style->progressStart($totalResults);
+
+        $fieldIdentifierInput = $this->input->getOption('field-identifier');
+        if ($fieldIdentifierInput === null) {
+            $fieldIdentifier = $this->configResolver->getParameter('tag_command_default_field_identifier', 'ngsite')[0];
+        } else {
+            $fieldIdentifier = $fieldIdentifierInput;
+        }
+
+        for ($offset = 0; $offset < $totalResults; $offset += $batchSize) {
+            $query->offset = $offset;
+            $query->limit = $batchSize;
+            $searchResults = $this->repository->getSearchService()->findContent($query);
+
+            $this->repository->beginTransaction();
```

Comment:
Why do you need a transaction for every batch? Every content needs its own transaction.

Reply:
That one was on me - won't that slow down everything too much? In my experience, it is much better to have a whole batch in the transaction, and commit once the whole batch is processed, rather than having one transaction per content.

Reply:
Maybe, but the only way to know for sure is to test it. This way, with one rollback, you effectively skip the entire batch of the content you need to tag, and getting the script to run again to tag only those content items means going through the ENTIRE set AGAIN, since you can't specify the content IDs, but only parent location ID. To me, it's a big downside.

Besides, it's a script, who cares? It's designed to run for extended periods of time.

Reply:
@Gintron Please move the transaction to content item level and not batch level.

Reply:
solved in b6964393